server/src/main/java/org/candlepin/util/DERUtil.java

```diff
@@ -0,0 +1,320 @@
+
+            length = 0;
+            for (int i = 0; i < size; i++) {
+                int next = readAndTrack(s, count);
+
+                if (next < 0) {
+                    throw new EOFException("EOF found reading length");
+                }
+
+                length = (length << 8) + next;
+            }
+
+            if (length < 0) {
+                throw new IOException("corrupted stream - negative length found");
+            }
+        }
+
+        return length;
+    }
+
+    public static void writeLength(OutputStream out, int length) throws IOException {
+        writeLength(out, length, null);
+    }
+
+    /**
+     * Write an integer as a DER encoded definite length.
+     *
+     * @param out
+     * @param length
+     * @throws IOException if something goes wrong
+     */
+    public static void writeLength(OutputStream out, int length, Signer signer) throws IOException {
+        if (length > 127) {
+            int size = 1;
+            int val = length;
+
+            while ((val >>>= 8) != 0) {
+                size++;
+            }
+
+            byte b = (byte) (size | 0x80);
+            out.write(b);
+            if (signer != null) {
+                signer.update(b);
+            }
+
+            for (int i = (size - 1) * 8; i >= 0; i -= 8) {
+                b = (byte) (length >> i);
+                out.write(b);
+                if (signer != null) {
+                    signer.update(b);
+                }
+            }
+        }
+        else {
+            byte b = (byte) length;
+            out.write(b);
+            if (signer != null) {
+                signer.update(b);
+            }
+        }
+    }
+
+    public static void writeTag(OutputStream out, int tag, int tagNo) throws IOException {
+        writeTag(out, tag, tagNo, null);
+    }
+
+    public static void writeTag(OutputStream out, int tag, int tagNo, Signer signer) throws IOException {
+        int rebuiltTag = rebuildTag(tag, tagNo);
+        out.write(rebuiltTag);
+        if (signer != null) {
+            signer.update((byte) rebuiltTag);
+        }
+    }
+
+    public static void writeValue(OutputStream out, byte[] value) throws IOException {
+        writeValue(out, value, null);
+    }
+
+    public static void writeValue(OutputStream out, byte[] value, Signer signer) throws IOException {
+        out.write(value);
+        if (signer != null) {
+            signer.update(value, 0, value.length);
+        }
+    }
+
+    public static int rebuildTag(int tag, int tagNo) {
+        // Universal tags have zeroes as the first two bits.
+        if ((tag >>> 6) != 0) {
+            throw new RuntimeException("This class only supports universal ASN1 tags.");
+        }
+
+        return tag;
+    }
+
+    public static void writeDER(OutputStream out, byte[] der) throws IOException {
+        writeDER(out, der, null);
+    }
+
+    public static void writeDER(OutputStream out, byte[] der, Signer signer) throws IOException {
```

Comment:
> This seems to be the same as writeValue(). Why have two methods to do the same thing?

They are functionally the same, but semantically different.  `writeValue` is for writing the *V* portion of the TLV and `writeDER` is for writing the entire TLV.  That distinction is not discussed in any comments or otherwise noted.  I'll probably just have `writeDER` delegate to `writeValue` or vice versa to preserve the distinction.  Or perhaps just change them all to `writeBytes`.